src/xmppjs/HistoryManager.ts

```diff
@@ -0,0 +1,102 @@
+import { Element } from "@xmpp/xml";
+import { JID } from "@xmpp/jid";
+
+export interface IHistoryLimits {
+    maxChars?: number | undefined,
+    maxStanzas?: number | undefined,
+    seconds?: number | undefined,
+    since?: Date | undefined,
+}
+
+/**
+ * Abstraction of a history storage backend.
+ */
+interface IHistoryStorage {
+    // add a message to the history of a given room
+    addMessage: (chatName: string, message: Element, jid: JID) => Promise<void>;
+    // get the history of a room.  The storage should apply the limits that it
+    // wishes too, and remove the limits that it applied from the `limits`
+    // parameter.  The returned list of Elements must include the Delayed
+    // Delivery information.
+    getHistory: (chatName: string, limits: IHistoryLimits) => Promise<Element[]>;
+}
+
+// pad a number to be two digits long
+function pad2(n: number) {
+    return n < 10 ? "0" + n.toString() : n.toString();
+}
+
+/**
+ * Store room history in memory.
+ */
+export class MemoryStorage implements IHistoryStorage {
+    private history: Map<string, Element[]>;
+    constructor(public maxHistory: number) {
+        this.history = new Map();
+    }
+
+    async addMessage(chatName: string, message: Element, jid: JID): Promise<void> {
+        if (!this.history.has(chatName)) {
+            this.history.set(chatName, []);
+        }
+        const currRoomHistory = this.history.get(chatName);
+
+        // shallow-copy the message, and add the timestamp
+        const copiedMessage = new Element(message.name, message.attrs);
+        copiedMessage.append(message.children as Element[]);
+        copiedMessage.attr("from", jid.toString());
+        const now = new Date();
+        copiedMessage.append(new Element("delay", {
+            xmlns: "urn:xmpp:delay",
+            from: chatName,
+            // based on the polyfill at
+            // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString
+            stamp: now.getUTCFullYear() +
```

Comment:
We're using the TypeScript transpiler and JS2015 classes. I'm pretty sure `.toISOString` is supported as well.

A real polyfill would check if `Date.prototype.toISOString` exists and, if not, create it.

Reply:
It's not doing the same as `.toISOString`, though.  `.toISOString` includes the milliseconds.  I guess I should note that.

Reply:
Turns out that XMPP will accept milliseconds, so switching to just using `.toISOString`